src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpTraceHelper.cs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Security;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpTraceHelper
+    {
+        private const string WinHtpTraceEnvironmentVariable = "WINHTTPHANDLER_TRACE";
+        private static bool s_TraceEnabled;
+
+        static WinHttpTraceHelper()
+        {
+            string env;
+            try
+            {
+                env = Environment.GetEnvironmentVariable(WinHtpTraceEnvironmentVariable);
+            }
+            catch (SecurityException)
+            {
+                env = null;
+            }
+            
+            if (!string.IsNullOrEmpty(env))
+            {
+                s_TraceEnabled = true;
+            }
+            else
+            {
+                s_TraceEnabled = false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsTraceEnabled()
+        {
+            if (s_TraceEnabled)
+            {
+                return true;
+            }
+            
+            return false;
+        }
+
+        public static void TraceStatusCallback(string message, IntPtr handle, bool invokeCallback, uint status)
+        {
+            if (!IsTraceEnabled())
+            {
+                return;
+            }
+
+            Debug.WriteLine(
+                "{0}: handle=0x{1:X}, {2}, {3}",
+                message,
+                handle,
+                invokeCallback ? "processing" : "skipping",
+                GetStringFromInternetStatus(status));
+        }
+
+        public static void TraceAsyncError(string message, Interop.WinHttp.WINHTTP_ASYNC_RESULT asyncResult)
+        {
+            if (!IsTraceEnabled())
+            {
+                return;
+            }
+
+            uint apiIndex = (uint)asyncResult.dwResult.ToInt32();
+            uint error = asyncResult.dwError;
+
+            Debug.WriteLine(
+                "{0}: api={1}, error={2}({3}) \"{4}\"",
+                message,
+                GetNameFromApiIndex(apiIndex),
+                GetNameFromError(error),
+                error,
+                WinHttpException.GetErrorMessage((int)error));
+        }
+
+        public static void Trace()
+        {
+            if (!IsTraceEnabled())
+            {
+                return;
+            }
+        }
+
+        private static string GetNameFromApiIndex(uint index)
+        {
+            switch (index)
+            {
+                case Interop.WinHttp.API_RECEIVE_RESPONSE:
```

Comment:
@bartonjs I'm not totally against changing it from `uint` to `enum`.  But the Interop file `winhttp_types.cs` has been in the repo a while now and was reviewed previously.

This issue is sort of orthogonal to this PR. But if we think overall performance might benefit as we increase the tracing support to include translating from uint to string, I would be ok with changing it to `enum`.